### ISSUE TITLE
BUG: Variables from variants not populated in build scripts

### ISSUE THREAD

**Issue by h-vetinari** [prefix-dev/rattler-build#1056]
While trying to [build](https://github.com/conda-forge/zlib-feedstock/pull/84) zlib for emscripten, I'm setting the following in CBC
```yaml
cross_target_platform:  # [linux64]
  - linux-64            # [linux64]
  - emscripten-wasm32   # [linux64]
```
and have a branch in the build script à la:
```bash
if [[ ${cross_target_platform} == emscripten-* ]]; then
    # ...
fi
```

However, when building this with rattler-build, the variables end up not being set, i.e. the log shows
```
 │ │ + [[ '' == emscripten-* ]]
```
where the `''` part should not be empty.

That's despite this being picked up for the various outputs - but perhaps not for `cache:`? 🤔 
```
 │ Build variant: zlib-1.3.1-hcf6bdf7_2
 │ 
 │ ╭───────────────────────┬───────────────────╮
 │ │ Variant               ┆ Version           │
 │ ╞═══════════════════════╪═══════════════════╡
 │ │ c_compiler            ┆ gcc               │
 │ │ c_compiler_version    ┆ 13                │
 │ │ c_stdlib              ┆ sysroot           │
 │ │ c_stdlib_version      ┆ 2.17              │
 │ │ channel_targets       ┆ conda-forge main  │
 │ │ cross_target_platform ┆ emscripten-wasm32 │     <--- see here
 │ │ libzlib               ┆ 1.3.1 h7eb9a27_2  │
 │ │ target_platform       ┆ linux-64          │
 │ ╰───────────────────────┴───────────────────╯
```

**Comment by wolfv** [comment]
Yes, you'll have to use the variable in the cache inside the YAML somehow. I think there is a force use variables option you could try, for example.

**Comment by wolfv** [comment]
Or you can pass it as an env var in the script context. We don't scan build scripts for used variables.

**Comment by h-vetinari** [comment]
> We don't scan build scripts for used variables.

It's not a question of scanning the build scripts, which I agree is not necessary, or even a good idea. But the variables coming from the variant config should be _defined_ during the build scripts, unless there's a very strong reason not to do this...?

A huge amount of recipes depend on this, it'll be really painful to work around this IMO.

**Comment by wolfv** [comment]
It's not the way conda build works. Variables have to be used to influence the variant.

**Comment by wolfv** [comment]
You can have a very long variant configuration file but only a subset will apply to your recipe. 

Conda forge just pre processes all of this in conda-smithy.

**Comment by h-vetinari** [comment]
> Variables have to be used to influence the variant.

I understand that; but it's enough to use it in the recipe

> It's not the way conda build works. 

conda-build _does_ populate variants (picked up from the recipe) also for the build script. Which is my point, because a lot of recipes rely on things being defined in the `.ci_support/*.yaml` files (however they get picked up by conda-build as "used") to be defined during the build script.

**Comment by h-vetinari** [comment]
In the case of https://github.com/conda-forge/zlib-feedstock/pull/84, a rerender with rattler-build populates `cross_target_plaform` in `.ci_support/linux_64_cross_target_platformemscripten-wasm32.yaml` as expected (because it appears in the recipe; i.e. scanning the build script is not necessary).

The only thing to do AFAIU is to pre-populate the environment variables present during the build scripts with whatever is in the respective variant config -- conda-build does it, recipes rely on it, and so it'd be mega-painful to transition recipes at scale if rattler-build won't do the same.

**Comment by wolfv** [comment]
But that would be wrong since we don't create a variant for it...

You need to __use__ the variant key in the output to make it appear in the build script. Otherwise the behavior would be super broken.

**Comment by wolfv** [comment]
For example, you could have multiple outputs. One depends on python and one does not. 

Only the one that depends on python will build the python variant packages. The python variant value will not be set during the build for the non-python output (even though it appears in the variant config). The value would be garbage anyways.

Does that make sense?

**Comment by wolfv** [comment]
The most obvious way would be to do:

```yaml
build:
  script:
     file: foo.sh
     env:
       CROSS_TARGET_PLATFORM: ${{ cross_target_platform }}
```

In the appropriate output.

The other way would be to use 

```yaml
build:
  variant:
     use_keys: ["cross_target_platform"]
```

In conda-build this probably works because they do analyze the `build.sh/bld.bat` file with a regex.

**Comment by h-vetinari** [comment]
So first I did:
```diff
    script:
      # rattler-build figures out the {.sh,.bat} extensions
      - build_global
+     # see https://github.com/prefix-dev/rattler-build/issues/1056
+     env:
+       CROSS_TARGET_PLATFORM: ${{ cross_target_platform }}
```
but that doesn't work because then one needs to shift 
```yaml
    script:
      - build_global
```
to
```yaml
    script:
      file: build_global
```

This is pretty bad dev experience IMO (I had two other failed attempts in between, because of invisible list-vs-scalar expectations). Why not make `file:` mandatory rather than let people fall into a trap once they need to add an `env:` or `secrets:`?

PS. It does work in the end, but I don't think it's a feasible to do a migration with this.

**Comment by h-vetinari** [comment]
> What you typed is also not valid YAML though (mixing list & map)

Sure (and I did try with `- env` too), but most parts of the current recipe infrastructure are built to work additively in terms of lines. Having to change the way the script-name is passed because we're adding `env:` is surprising/annoying.

**Comment by wolfv** [comment]
OK, I think I initially understood your request slightly wrongly. The bug you have discovered is about making the variant key/values available as environment vars in the build script. That is something that seems to be broken in rattler-build indeed. I'll double check if that is also a problem for a regular single-output build.

We will definitely fix that! :)

**Comment by h-vetinari** [comment]
> The bug you have discovered is about making the variant key/values available as environment vars in the build script.

👍

Sorry I didn't manage to communicate this more clearly.

**Comment by wolfv** [comment]
Fixed by #1088
